pennylane/argmap.py

```diff
@@ -0,0 +1,251 @@
+        if like is not None:
+            if not isinstance(like, ArgMap):
+                raise ArgMapError("Trying to inherit properties from non-ArgMap object.")
+            single_arg, single_object = like.single_arg, like.single_object
+        self.single_arg = single_arg
+        self.single_object = single_object
+        _data = self._preprocess_data(data)
+        super().__init__(_data)
+        self.consistency_check()
+
+    def _preprocess_data(self, data):
+        if data is None:
+            data = {}
+        if self.single_object:
+            return {(None, None): data}
+
+        if not isinstance(data, dict):
+            try:
+                data = dict(data)
+            except (ValueError, TypeError) as e:
+                raise ArgMapError(
+                    f"The input could not be interpreted as dictionary; input:\n{data}"
+                ) from e
+
+        return {_interpret_key(key, self.single_arg): val for key, val in data.items()}
+
+    def __getitem__(self, key):
+        key = _interpret_key(key, self.single_arg)
+        return super().__getitem__(key)
+
+    def __setitem__(self, key, value):
+        key = _interpret_key(key, self.single_arg)
+        return super().__setitem__(key, value)
+
+    def setdefault(self, key, default=None):
+        key = _interpret_key(key, self.single_arg)
+        return super().setdefault(key, default)
+
+    def __eq__(self, other):
+        if isinstance(other, ArgMap):
+            if (self.single_arg + other.single_arg) % 2:
+                return False
+            if (self.single_object + other.single_object) % 2:
+                return False
+        return super().__eq__(other)
+
+    def __ne__(self, other):
+        if isinstance(other, ArgMap):
+            if (self.single_arg + other.single_arg) % 2:
+                return True
+            if (self.single_object + other.single_object) % 2:
+                return True
+        return super().__ne__(other)
+
+    def get(self, key, default=None):
+        key = _interpret_key(key, self.single_arg)
+        return super().get(key, default)
+
+    def consistency_check(self, check_values=False):
+        r"""Check the stored keys and optionally values for consistency."""
+        if self.single_object:
+            self._check_single_object()
+            if check_values:
+                self._check_values()
+            return
+        if any((key == (None, None) for key in self)):
+            raise ArgMapError(
+                "The key (None, None) indicates single object but ArgMap.single_object=False."
+            )
+        if self.single_arg:
+            self._check_single_arg()
+        else:
+            self._check_multiple_args()
+        if check_values:
+            self._check_values()
+
+    def _check_values(self):
+        types = [(key, type(val)) for key, val in self.items()]
+        if not all((_type[1] == types[0][1] for _type in types)):
```

Comment:
```suggestion
        if not all(_type[1] == types[0][1] for _type in types):
```

Reply:
Should it be `is` instead of `==`?

Reply:
Wait, one can skip the other pair of parentheses :sweat_smile: Thanks!
And yes, `is` is the appropriate choice! :+1: